### DISCUSSION TITLE
ViewModelLocationProvider is NOT working when view is not in Views folder

### DISCUSSION THREAD

**Discussion by ldgragg** [PrismLibrary/Prism#2643] [WPF]
I have tried everything I can think of and find online to make this connection but nothing will connect up the two and create the VM.

I have a View in a menu folder and the corresponding VM is in the ViewModels folder ....

I tried the method BL used in his Outlook series, putting the ViewModelLocationProvider.Register in the ApplicationsModule.cs file at the root of the module ..... didn't work.

I tried all of the different syntaxes that the Register method provides .... didn't work.

In the attached file I am trying the override of the ConfigureVieModelLocator() and adding the Register there ... doesn't work.

Any help would be very much appreciated.

[ApplicationsGroup.txt](https://github.com/PrismLibrary/Prism/files/7746533/ApplicationsGroup.txt)
[App.xaml.txt](https://github.com/PrismLibrary/Prism/files/7746553/App.xaml.txt)

**Comment by dansiegel** [comment]
Just to be clear, Prism doesn't care about the path on disk as much as the namespace. By convention the ViewModelLocator only works when the ViewModel is in the `ViewModels` namespace. 

You have a couple of options.
1. The first option, which we recommend to reduce the impact of reflection lookups is to register your Views with the ViewModel mapping:

```cs
containerRegistry.RegisterForNavigation<ViewA, ViewAViewModel>();
```

2. The other option is to specify an override of the default lookup behavior shown below.

```cs
ViewModelLocationProvider.SetDefaultViewTypeToViewModelTypeResolver(type => {
  // Your logic here..
});
```
https://github.com/PrismLibrary/Prism/blob/7f0b1680bbe754da790274f80851265f808d9bbf/src/Prism.Core/Mvvm/ViewModelLocationProvider.cs#L46-L54

**Reply by ldgragg** [reply]
Dan,

Thanks for the reply ..

I understand that and believe me have tried everything...... The VM is in the VM folder / VM namespace but the View is in the Menu folder / Menu namespace.

I have tried the RegisterForNavigation ........ it did not work.  

All of my V/VM' are linking up and working perfect except for this one because it is in the Menu folder/namespace.

According to all the info i have read and the videos I have watched the ViewModelLocationProvider.Register<ApplicationsGroup, ApplicationsGroupViewModel>();  should work but it does not.

The only thing I have not tried is your option 2 from above ..... where can i see an example of this?

**Reply by dansiegel** [reply]
can you provide a reproduction of your issue? It's much easier to give you feedback about what may be wrong if we have a sample that reproduces the issue.

**Comment by ldgragg** [comment]
Ok, so I have dug much much deeper and discovered that it appears for AutoWireViewModel to work the control must be a FrameworkElement descendent and the NavBarGroup component in the DevExpress library is descended from FrameworkContentElement.  So I think this is why the AutoWire isnt working.  Does anyone have any work around ideas?

**Reply by bakerhillpins** [reply]
@ldgragg  I might be a bit late here.  I'm taking a guess at what you're trying to do here. NavBarGroup sounds like you're trying to bind a collection of objects to an ItemsSource and the NavBarGroup is a sub-group in the collection? (Ok looked up the DevExpress control, and that seems reasonable).

I'll guess that you've already created your own RegionAdapter for the NavBarControl since it doesn't appear to fit into any of the out of the box WPF adapters (It inherits directly from System.Windows.Controls.Control). This should allow you the ability to change your design to discover/inject in a supported View (FrameworkElement descendent) and then go with a DataTemplateSelector like approach where you wrap that injected view into a NavBarGroup when it's added to the NavBarControl.  

Basically, make your view the Grid, rather than the NavBarGroup. Then when the View/Grid is injected into the region use a DataTemplate to wrap the Grid in a NavBarControl. In the template you can promote the View's DataContext value up to the parent NavBarControl.DataContext via a binding or make sure your NavBarGroup bindings correctly reference the View's DataContext (ViewModel) data in binding statements.  

Create the template as a resource with a key, access the template with the key, create the wrapper by calling the DataTemplate.LoadContent() method (Results in a NavBarControl) and set the NavBarControl.Content property to the view.

Assuming you based your Adapter on the SelectorRegionAdapter, you'd probably make these changes in the SelectorItemsSourceSyncBehavior.Views_CollectionChanged event handler. You'd need to update any logic that compares/looks up a view as well.
